File: website/APIs/nest/my_nest.py
import os
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
import pickle
from googleapiclient.discovery import build


# The CLIENT_ID and CLIENT_SECRET you got from the Google Cloud Console
CLIENT_ID = '986207922032-26pg2rnsrbjf5f1lk224h42omp4u9ljt.apps.googleusercontent.com'
CLIENT_SECRET = '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'

# The scope for the Smart Device Management API
SCOPES = ['https://www.googleapis.com/auth/sdm.service']

creds = None

# The file token.pickle stores the user's access and refresh tokens, and is
# created automatically when the authorization flow completes for the first time.
if os.path.exists('token.pickle'):
    with open('token.pickle', 'rb') as token:
        creds = pickle.load(token)

# If there are no (valid) credentials available, let the user log in.
if not creds or not creds.valid:
    if creds and creds.expired and creds.refresh_token:
        creds.refresh(Request())
    else:
        flow = InstalledAppFlow.from_client_config(
            {"web": {"client_id": CLIENT_ID, "client_secret": CLIENT_SECRET, "auth_uri": "https://accounts.google.com/o/oauth2/auth", "token_uri": "https://oauth2.googleapis.com/token"}},
            scopes=SCOPES
        )
        creds = flow.run_local_server(port=41567)
    # Save the credentials for the next run
    with open('token.pickle', 'wb') as token:
        pickle.dump(creds, token)

# Build the service
service = build('smartdevicemanagement', 'v1', credentials=creds)

# Replace 'enterprises/project-id' with your Project ID
request = service.enterprises().devices().list(parent='enterprises/0fb5e183-9cef-4050-b3fe-8184d6ff5951')

# Execute the request
response = request.execute()

# Assuming devices is your list of devices:
devices = response.get('devices', [])
for device in devices:
    if device.get('type') == 'sdm.devices.types.THERMOSTAT':
        device_name = device.get('name')

        if device_name:
            try:
                # Get details of the device
                request = service.enterprises().devices().get(name=device_name)
                response = request.execute()

                # Extract the current temperature in Celsius, if available
                traits = response.get('traits', {})
                temperature_traits = traits.get('sdm.devices.traits.Temperature')
                if temperature_traits:
                    current_temperature_celsius = temperature_traits.get('ambientTemperatureCelsius')
                    if current_temperature_celsius is not None:
                        print(f"Current temperature: {current_temperature_celsius}°C")
                        
                    else:
                        print("Current temperature data is not available.")
                else:
                    print("Temperature traits not found for this device.")
            except Exception as e:
                print(f"Error occurred while fetching device details: {str(e)}")
        else:
            print("Device name is not available.")
    else:
        print("No thermostat found.")

File: backend/apps/nest/my_nest.js
const fs = require('fs');
const { google } = require('googleapis');
const express = require('express');
const open = import('open');
const OAuth2 = google.auth.OAuth2;

const CLIENT_ID = '986207922032-26pg2rnsrbjf5f1lk224h42omp4u9ljt.apps.googleusercontent.com';
const CLIENT_SECRET = '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';
const SCOPES = ['https://www.googleapis.com/auth/sdm.service'];
const TOKEN_PATH = 'token.json';

let credentials = null;
if (fs.existsSync(TOKEN_PATH)) {
    const token = fs.readFileSync(TOKEN_PATH);
    credentials = JSON.parse(token);
}

const oauth2Client = new OAuth2(
    CLIENT_ID,
    CLIENT_SECRET,
    'http://localhost:41567'
);

const service = google.smartdevicemanagement({
    version: 'v1',
    auth: oauth2Client,
});

function fetchDevices() {
    const devicesListRequest = service.enterprises.devices.list({
        parent: 'enterprises/0fb5e183-9cef-4050-b3fe-8184d6ff5951',
    });

    devicesListRequest.then(response => {
        const devices = response.data.devices || [];
        devices.forEach(device => {
            if (device.type === 'sdm.devices.types.THERMOSTAT') {
                const deviceName = device.name;
                if (deviceName) {
                    service.enterprises.devices.get({name: deviceName}).then(response => {
                        const traits = response.data.traits || {};
                        const temperatureTraits = traits['sdm.devices.traits.Temperature'];
                        if (temperatureTraits) {
                            const currentTemperatureCelsius = temperatureTraits.ambientTemperatureCelsius;
                            if (currentTemperatureCelsius !== undefined) {
                                console.log(`Current temperature: ${currentTemperatureCelsius}°C`);
                            } else {
                                console.log("Current temperature data is not available.");
                            }
                        } else {
                            console.log("Temperature traits not found for this device.");
                        }
                    }).catch(err => {
                        console.error(`Error occurred while fetching device details: ${err}`);
                    });
                } else {
                    console.log("Device name is not available.");
                }
            } else {
                console.log("No thermostat found.");
            }
        });
    }).catch(err => {
        console.error(`Error occurred while fetching devices: ${err}`);
    });
}

if (!credentials) {
    const authUrl = oauth2Client.generateAuthUrl({
        access_type: 'offline',
        scope: SCOPES,
    });

    console.log('Authorize this app by visiting this url:', authUrl);

    const app = express();

    app.get('/', (req, res) => {
        const code = req.query.code;
        oauth2Client.getToken(code, (err, token) => {
            if (err) {
                console.error('Failed to retrieve access token', err);
                return res.status(500).send(err);
            }

            oauth2Client.setCredentials(token);
            fs.writeFileSync(TOKEN_PATH, JSON.stringify(token));

            console.log('Token stored to', TOKEN_PATH);
            res.send('Success! Please return to the application.');

            fetchDevices();
        });
    });

    app.listen(41567, async () => {
        console.log(`App listening at http://localhost:41567`);
        const openModule = await open;
        openModule.default(authUrl);
    });
} else {
    oauth2Client.setCredentials(credentials);
    fetchDevices();
}
